code/game/objects/items/RCD.dm

```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+				icon = "[design["title"]]"
+			else if(sub_category == "Glass AirLocks")
+				icon = "[design["title"]]-Glass"
+			else
+				icon = design["icon"]
+			icon = replacetext(icon," ","-") //css class names cannot have spaces in them
+
+			designs += list(list("design_id" = i, "title" = design["title"], "icon" = icon))
+		data["categories"] += list(list("cat_name" = sub_category, "designs" = designs))
+
+	//merge airlock_electronics ui data with this
+	var/list/airlock_data = airlock_electronics.ui_data(user)
+	for(var/key in airlock_data)
+		data[key] = airlock_data[key]
+
+	return data
+
+/obj/item/construction/rcd/ui_act(action, params)
+	..()
+	if(.)
 		return
-	switch(choice)
-		if("Floors & Walls")
-			construction_mode = RCD_FLOORWALL
-		if("Airlock")
-			construction_mode = RCD_AIRLOCK
-		if("Grilles & Windows")
-			construction_mode = RCD_WINDOWGRILLE
-		if("Machine Frames")
-			construction_mode = RCD_MACHINE
-		if("Furnishing")
-			construction_mode = RCD_FURNISHING
-		if("Computer Frames")
-			construction_mode = RCD_COMPUTER
-			change_computer_dir(user)
-			return
-		if("Change Access")
-			airlock_electronics.ui_interact(user)
-			return
-		if("Change Airlock Type")
-			change_airlock_setting(user)
-			return
-		if("Change Window Glass")
-			toggle_window_glass(user)
-			return
-		if("Change Window Size")
-			toggle_window_size(user)
-			return
-		if("Change Furnishing Type")
-			change_furnishing_type(user)
-			return
-		if("Silo Link")
-			toggle_silo_link(user)
-			return
+
+	switch(action)
+		if("root_category")
+			var/new_root = params["root_category"]
+			if(root_categories[new_root] != null || new_root == "AirLock Access") //is a valid category
+				root_category = new_root
+
+		if("design")
+			var/category_name = params["category"]
+			var/index = params["index"]
+
+			var/list/root = root_categories[root_category]
+			if(root == null) //not a valid root
+				return TRUE
+			var/list/category = root[category_name]
+			if(category == null) //not a valid category
+				return TRUE
+			var/list/design = category[index]
+			if(design == null) //not a valid design
+				return TRUE
+
+			design_category = category_name
+			design_title = design["title"]
+
+			if(category_name == "Structures")
+				construction_mode = design["construction_mode"]
+				if(design["window_type"] != null)
+					window_type = design["window_type"]
+				if(design["window_glass"] != null)
+					window_glass = design["window_glass"]
+				if(design["window_size"] != null)
+					window_size = design["window_size"]
+			else if(category_name == "Machines")
+				construction_mode = design["construction_mode"]
+				if(design["computer_dir"] != null)
+					computer_dir = design["computer_dir"]
+			else if(category_name == "Furniture")
+				construction_mode = RCD_FURNISHING
+				furnish_type = design["furnish_type"]
+				furnish_cost = design["furnish_cost"]
+				furnish_delay = design["furnish_delay"]
+
+			if(root_category == "AirLocks")
```

Comment:
Also why are these all "AirLocks" and not "Airlocks"?

Reply:
I'll change that